packages/near-api-js/src/providers/wallet-rpc-provider.ts

```diff
@@ -0,0 +1,473 @@
+/**
+ * @module
+ * @description
+ * This module contains the {@link JsonRpcProvider} client class
+ * which can be used to interact with the [NEAR RPC API](https://docs.near.org/api/rpc/introduction).
+ * @see {@link providers/provider | providers} for a list of request and response types
+ */
+import {
+    AccessKeyWithPublicKey,
+    Provider,
+    FinalExecutionOutcome,
+    NodeStatusResult,
+    BlockId,
+    BlockReference,
+    BlockResult,
+    BlockChangeResult,
+    ChangeResult,
+    ChunkId,
+    ChunkResult,
+    EpochValidatorInfo,
+    NearProtocolConfig,
+    LightClientProof,
+    LightClientProofRequest,
+    GasPrice,
+    QueryResponseKind
+} from './provider';
+import { TypedError, ErrorContext } from '../utils/errors';
+import { baseEncode } from 'borsh';
+import exponentialBackoff from '../utils/exponential-backoff';
+import { parseRpcError } from '../utils/rpc_errors';
+import { SignedTransaction, Transaction } from '../transaction';
+
+/** @hidden */
+export { TypedError, ErrorContext };
+
+// Default number of retries before giving up on a request.
+const REQUEST_RETRY_NUMBER = 12;
+
+// Default wait until next retry in millis.
+const REQUEST_RETRY_WAIT = 500;
+
+// Exponential back off for waiting to retry.
+const REQUEST_RETRY_WAIT_BACKOFF = 1.5;
+
+/// Keep ids unique across all connections.
+let _nextId = 123;
+
+export interface RequestParams {
+    jsonrpc: '2.0';
+    id: number;
+    method: string;
+    params: object;
+}
+
+export interface WalletProvider {
+    getAccount: () => Promise<string>;
+    request: (args: RequestParams) => Promise<any>;
+    on: (message: string, listener: (...args: any[]) => void ) => void;
+}
+
+/**
+ * Client class to interact with the [NEAR RPC API](https://docs.near.org/api/rpc/introduction).
+ * @see [https://github.com/near/nearcore/tree/master/chain/jsonrpc](https://github.com/near/nearcore/tree/master/chain/jsonrpc)
+ */
+export class WalletRpcProvider extends Provider {
+    /** @hidden */
+    private _provider: WalletProvider | null = null;
+
+    private _network = '';
+    private _account = '';
+    private _pubKeys: string[] = [];
+
+    /**
+     * get chain id (near or testnet)
+     */
+    get network(): string {
+        return this._network;
+    }
+
+    /**
+     * get selected account id
+     */
+    get account(): string {
+        return this._account;
+    }
+
+
+    /**
+     * get access key of selected account id
+     */
+    get pubKeys(): string[] {
+        return this._pubKeys;
+    }
+
+    /**
+     * get type of provider
+     */
+    get isWalletProvider(): boolean {
+        return true;
+    }
+
+    /**
+     * @param provider WalletProvider
+     */
+    constructor(provider: WalletProvider) {
+        super();
+        this._provider = provider;
+        this._provider.on('chainChanged', this.updateNetwork.bind(this));
+        this._provider.on('accountsChanged', this.updateAccount.bind(this));
+        this.init();
```

Comment:
Firing and forgetting an async method in the constructor here is problematic:

1. Any rejections will be unhandled
2. There is no way of knowing when this method has completed and thus when the object has been initialized

This initialization must occur _outside_ of this class or as an explicit secondary step after instantiation for behavior dependent on `init`'s resolution to be reliable at all.